### PR TITLE
Kierangould/12hourtimestamp

### PR DESCRIPTION
**Pull request by Kieran-Gould**
Added support for 12 hour timestamps. Option for 12/24hr is pulled from UserSettings.

### CODE REVIEW

File: src/DateUtils.js [hunk 1]
```diff
@@ -16,39 +16,51 @@ limitations under the License.
...
     return (n < 10 ? '0' : '') + n;
 }
 
+function twentyfourhour(date) {
+  let hours = date.getHours();
+  let minutes = date.getMinutes();
+  let ampm = hours >= 12 ? 'PM' : 'AM';
+  hours = hours % 12;
+  hours = hours ? hours : 12;
+  minutes = minutes < 10 ? '0'+minutes : minutes;
```
Comment (ara4n) [inline src/DateUtils.js]: i suggest just using the pad(n) function defined above :)

File: src/DateUtils.js [hunk 2]
```diff
@@ -16,39 +16,51 @@ limitations under the License.
...
+import UserSettingsStore from './UserSettingsStore';
+const days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
+const months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
+let time;
 
 function pad(n) {
     return (n < 10 ? '0' : '') + n;
 }
 
+function twentyfourhour(date) {
```
Comment (ara4n) [inline src/DateUtils.js]: please indent consistently as per the code-style (i.e. 4 spaces).  The function probably should be `twentyFourHour()` too for consistent casing.

File: src/components/structures/UserSettings.js [hunk 3]
```diff
@@ -65,7 +65,6 @@ const SETTINGS_LABELS = [
         id: 'dontSendTypingNotifications',
         label: "Don't send typing notifications",
     },
-/*
```
Comment (ara4n) [inline src/components/structures/UserSettings.js]: yay! \o/

File: src/components/views/rooms/EventTile.js [hunk 4]
```diff
@@ -479,24 +481,31 @@ module.exports = WithMatrixClient(React.createClass({
         );
 
         var e2e;
+        let e2e_style;
```
Comment (ara4n) [inline src/components/views/rooms/EventTile.js]: Hm. Changing the whole CSS to make room for 24h timestamps feels a bit worrying, although I see why you're doing it...

A nicer bet here would be to just put the `mx_EventTile_12h` class onto the EventTile itself, and then in the CSS have a selector like:

```
.mx_EventTile_12h .mx_EventTile_e2eIcon {
    padding-left: 5px;
}

.mx_EventTile_12h .mx_EventTile_line {
    padding-left: 100px;
}
```

...or whatever the tweaks are.  And then you don't need all these overrides of the e2e_style, which are quite confusing at first as it makes it sound like the e2e icon is somehow very explicitly being changed by the timestamp(!) rather than just for layout purposes.

TL;DR: cascading stylesheets cascade; let's make the most of that :)

**Comment by ara4n** [comment]
tests are failing - my guess (untested) is that you need `mkStubRoom()` in test/test-utils.js to return a valid event for `getAccountData()`, like `createTestClient()` does.  Atm it returns null, hence the NPE.

I'm also slightly having second thoughts about whether EventTile should be using a prop for this (which is set by the room itself) rather than querying the js-sdk from the depths of its render() method.  If it was set by the room, then we'd have the advantage of being able to listen for `Room.accountData` events and update the prop when that happens, thus responding to synced settings changes in realtime rather than waiting for the page to be reloaded...

**Comment by ara4n** [comment]
lgtm! sorry for slow review; we're drowning.
